### PR TITLE
split_controller: fix the bug that may cause the stats_monitor to fall into a dead loop

### LINKED ISSUE #12416: TiKV Short-term memory usage continues to rise
### What version of TiKV are you using?

v6.0.0
2b4f7d57af9ca3e21bc07b91a9cfc466ed87734f

### What did happened?

image
image

Suspect the same problem as this one：https://github.com/tikv/tikv/issues/12229

### PR DESCRIPTION
**Pull request by JmPotato**
Signed-off-by: JmPotato <ghzpotato@gmail.com>

### What is changed and how it works?

Issue Number: Close #12416

What's Changed:

```commit-message
split_controller: fix the bug that may cause the stats_monitor to fall into a dead loop
```

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

```release-note
Fix a bug that causes the TiKV thread to fall into a dead loop and then lead to OOM.
```

### CODE REVIEW

**Review by lhy1024 — COMMENTED** [review]
LGTM. But do we need to revert max_retry to avoid other similar problems?

**Comment by nolouch** [comment]
cc @hundundm

**Review by nolouch — COMMENTED** [review]
Can we try to add some fuzz test for the function?

**Comment by JmPotato** [comment]
> Can we try to add some fuzz test for the function?

Yes, I'm working on it.

**Comment by JmPotato** [comment]
> LGTM. But do we need to revert max_retry to avoid other similar problems?

After this PR, I believe there is no need to add this mechanism.

File: components/raftstore/src/store/worker/split_controller.rs [hunk 1]
```diff
@@ -81,11 +81,15 @@ where
     F: Fn(&mut T) -> &mut Vec<KeyRange>,
 {
     let mut sampled_key_ranges = vec![];
+    // Retain the non-empty key ranges.
+    // `key_ranges_provider` may return an empty key ranges vector, which will cause
+    // the later sampling to fall into a dead loop. So we need to filter it out here.
+    key_ranges_providers
+        .retain_mut(|key_ranges_provider| !key_ranges_getter(key_ranges_provider).is_empty());
```
Comment (AndreMouche) [inline components/raftstore/src/store/worker/split_controller.rs:88]: So it seems the key_ranges_providers won't too big so we run the function `key_ranges_getter` directly on item? If yes, Could we store the result of `key_range_getter(key_ranges_provider)`  and use it directly on L122?

Or If the `key_ranges_providers` maybe very big, we could consider to control  `while` loop instead?

File: components/raftstore/src/store/worker/split_controller.rs [hunk 2]
```diff
@@ -905,33 +933,54 @@ mod tests {
...
+                let all_key_ranges_num = *prefix_sum(key_ranges.iter(), Vec::len).last().unwrap();
+                assert_eq!(
+                    sampled_key_ranges.len(),
+                    std::cmp::min(sample_num, all_key_ranges_num)
+                );
+            }
+        }
+    }
+
+    fn check_sample_length_with_timeout(key_ranges: Vec<Vec<KeyRange>>, timeout: Duration) {
```
Comment (lhy1024) [inline components/raftstore/src/store/worker/split_controller.rs:949]: if we revert DEFAULT_MAX_SAMPLE_LOOP_COUNT, do we still need it?

**Review by nolouch — COMMENTED** [review]
LGTM

File: components/raftstore/src/store/worker/split_controller.rs [hunk 3]
```diff
@@ -121,6 +132,12 @@ where
             sampled_key_ranges.push(key_ranges.remove(j)); // Sampling without replacement
         }
     }
+    if sample_loop_count == 0 {
+        warn!("The number of sampled key ranges could be less than the sample_num, the sampling may fall into a dead loop before";
```
Comment (nolouch) [inline components/raftstore/src/store/worker/split_controller.rs:136]:
```suggestion
        warn!("the number of sampled key ranges could be less than the sample_num, the sampling may fall into a dead loop before";
```

**Review by nolouch — APPROVED** [review]

**Review by lhy1024 — APPROVED** [review]

**Review by AndreMouche — APPROVED** [review]
LGTM

**Pull request merged**